README.md
# Faketooth Flutter Plugin

The Faketooth Flutter Plugin allows you to emulate a Bluetooth Low Energy (BLE) device on iOS, macOS, watchOS, and tvOS targets from within your Flutter applications. This plugin leverages the [`faketooth`](https://github.com/rozd/faketooth) library for iOS/macOS/watchOS/tvOS, enabling you to create virtual peripherals with custom services, characteristics, descriptors, and values.

## Installation

To use the Faketooth Flutter Plugin, you need to add it as a dependency in your `pubspec.yaml` file:

```yaml
dependencies:
  faketooth_flutter: ^0.2.1
```

Then, run `flutter pub get` to fetch the plugin.

## Usage

Follow the steps below to start using the Faketooth Flutter Plugin:

1. Import the package in your Dart code:

```dart
import 'package:faketooth_flutter/faketooth_flutter.dart';
```

2. Set up the simulated peripherals by calling the `setSimulatedPeripherals` method on the `Faketooth` singleton instance. Pass in a list of `FaketoothPeripheral` objects to define the virtual BLE devices you want to emulate.

Here's an example that demonstrates the basic usage:

```dart
await Faketooth.shared.setSimulatedPeripherals([
  FaketoothPeripheral(
    identifier: 'E621E1F8-C36C-495A-93FC-0C247A3E6E5F',
    name: "TDP_SIMULATED",
    services: [
      FaketoothService(
        uuid: '0000180A-0000-1000-8000-00805F9B34FB',
        isPrimary: true,
        characteristics: [
          FaketoothCharacteristic(
            uuid: '2A29',
            properties: {
              FaketoothCharacteristicProperties.read,
              FaketoothCharacteristicProperties.notify,
            },
            valueProducer: () {
              return Future.value(Uint8List.fromList('Hello'.codeUnits));
            },
          ),
          // Add more characteristics and services as needed
        ],
      ),
      // Add more services as needed
    ],
  ),
  // Add more peripherals as needed
]);
```

In this example, we create a single virtual peripheral named "TDP_SIMULATED" with a primary service. The service contains a characteristic with read and notify properties. You can add more characteristics and services based on your requirements.

3. Build and run your Flutter application. Faketooth will simulate the BLE interface and make it possible to interact with your virtual peripherals as if they were real devices.

## Example

To see the Faketooth Flutter Plugin in action, you can check out the example provided in the [Faketooth Flutter](https://github.com/rozd/faketooth-flutter/tree/master/example) repository. It demonstrates how to set up and use the plugin within a Flutter project.

## Contributions and Support

Contributions to the Faketooth Flutter Plugin are welcome! If you encounter any issues, have questions, or would like to suggest improvements, please open an issue on the [GitHub repository](https://github.com/rozd/faketooth-flutter).

## License

The Faketooth Flutter Plugin is released under the MIT license. See the [LICENSE](https://github.com/rozd/faketooth-flutter/blob/main/LICENSE) file for more information.

---

Thank you for using the Faketooth Flutter Plugin! I hope this plugin simplifies your BLE testing and development workflows. If you find it helpful, please consider giving it a star on [GitHub](https://github.com/rozd/faketooth-fl


CHANGELOG.md
## 0.2.1
Updates iOS podspec

## 0.2.0
Fixed `setValueForDescriptor` handler. Added code documentation. 

## 0.1.2
Minor changes, updated documentation.

## 0.1.1
Null safety

## 0.0.25
Added value handlers for Characteristic and Descriptor classes.

## 0.0.24
Added possibility to specify characteristic and descriptor initial values.

## 0.0.23
Updated Faketooth iOS package with fixes for characteristic notification.

## 0.0.22
Fixed inaccessible FaketoothDescriptor's fields

## 0.0.21
Added support for setting delays

## 0.0.20
Fixed serialization for Characteristic's properties

## 0.0.19
Initial support for simulating Advertisement Data

## 0.0.18
Added support for notifying characteristic's value

## 0.0.9
Initial release

lib/faketooth_characteristic.dart
part of faketooth;

/// A class that represents a Bluetooth Low Energy (BLE) characteristic in Faketooth.
///
/// This class defines the properties and behavior of a BLE characteristic, including its UUID, properties, descriptors,
/// initial value, value producer, and value handler.
class FaketoothCharacteristic {
  /// The UUID of the characteristic.
  final String uuid;

  /// The set of properties associated with the characteristic.
  final Set<FaketoothCharacteristicProperties> properties;

  /// The list of descriptors associated with the characteristic.
  final List<FaketoothDescriptor>? descriptors;

  /// A function that produces the value of the characteristic.
  final Future<Uint8List> Function()? valueProducer;

  /// A function that handles the value changes of the characteristic.
  final void Function(Uint8List?)? valueHandler;

  /// The initial value of the characteristic.
  final Uint8List? initialValue;

  /// Creates a new instance of [FaketoothCharacteristic].
  ///
  /// The [uuid] and [properties] parameters are required, while the [descriptors], [initialValue], [valueProducer], and [valueHandler]
  /// parameters are optional.
  const FaketoothCharacteristic({
    required this.uuid,
    required this.properties,
    this.descriptors,
    this.initialValue,
    this.valueProducer,
    this.valueHandler,
  });

  /// Converts the characteristic to a map of arguments.
  ///
  /// This method is useful for passing the characteristic as arguments to the Faketooth simulator.
  ///
  /// Example usage:
  /// ```dart
  /// FaketoothCharacteristic characteristic = FaketoothCharacteristic(
  ///   uuid: '00002a37-0000-1000-8000-00805f9b34fb',
  ///   properties: {
  ///     FaketoothCharacteristicProperties.read,
  ///     FaketoothCharacteristicProperties.write,
  ///   },
  ///   descriptors: [
  ///     FaketoothDescriptor(
  ///       uuid: '00002902-0000-1000-8000-00805f9b34fb',
  ///       value: Uint8List.fromList([0x00, 0x00]),
  ///     ),
  ///   ],
  /// );
  ///
  /// Map<String, dynamic> arguments = characteristic.toArguments();
  /// print(arguments);
  /// // Output: {uuid: '00002a37-0000-1000-8000-00805f9b34fb', properties: ['read', 'write'], ...}
  /// ```
  Map<String, dynamic> toArguments() {
    return {
      'uuid'        : uuid,
      'properties'  : properties.map((e) => e.code).toList(),
      'descriptors' : descriptors?.map((e) => e.toArguments()).toList() ?? [],
      "initialValue": initialValue,
    };
  }
}

/// An enum that represents the properties of a Bluetooth Low Energy (BLE) characteristic in Faketooth.
///
/// The properties define the behavior and capabilities of a characteristic.
enum FaketoothCharacteristicProperties {
  /// The characteristic supports broadcasting.
  broadcast,

  /// The characteristic supports reading.
  read,

  /// The characteristic supports writing without response.
  writeWithoutResponse,

  /// The characteristic supports writing.
  write,

  /// The characteristic supports notifications.
  notify,

  /// The characteristic supports indications.
  indicate,

  /// The characteristic supports authenticated signed writes.
  authenticatedSignedWrites,

  /// The characteristic has extended properties.
  extendedProperties,

  /// The characteristic requires encryption for notifications.
  notifyEncryptionRequired,

  /// The characteristic requires encryption for indications.
  indicateEncryptionRequired,
}

extension FaketoothCharacteristicPropertiesCode on FaketoothCharacteristicProperties {
  /// Retrieves the code corresponding to the characteristic property.
  ///
  /// The code is used internally by Faketooth for configuration and communication.
  /// Returns `null` if the characteristic property is not recognized.
  String? get code {
    switch (this) {
      case FaketoothCharacteristicProperties.broadcast                  : return "broadcast";
      case FaketoothCharacteristicProperties.read                       : return "read";
      case FaketoothCharacteristicProperties.writeWithoutResponse       : return "writeWithoutResponse";
      case FaketoothCharacteristicProperties.write                      : return "write";
      case FaketoothCharacteristicProperties.notify                     : return "notify";
      case FaketoothCharacteristicProperties.indicate                   : return "indicate";
      case FaketoothCharacteristicProperties.authenticatedSignedWrites  : return "authenticatedSignedWrites";
      case FaketoothCharacteristicProperties.extendedProperties         : return "extendedProperties";
      case FaketoothCharacteristicProperties.notifyEncryptionRequired   : return "notifyEncryptionRequired";
      case FaketoothCharacteristicProperties.indicateEncryptionRequired : return "indicateEncryptionRequired";
      default: return null;
    }
  }
}